src/context.rs

```diff
@@ -1,88 +1,360 @@
+/// # use futures::future::{Future, ok};
+/// # use std::marker::PhantomData;
+/// #
+/// # struct MyItem;
+/// # fn do_something_with_my_item(item: &MyItem) {}
+/// #
+/// struct MiddlewareService<T, C> {
+///     inner: T,
+///     marker: PhantomData<C>,
+/// }
+///
+/// impl<T, C> hyper::server::Service for MiddlewareService<T, C>
+///     where
+///         C: Pop<MyItem>,
+///         T: hyper::server::Service<Request = (hyper::Request, C::Result)>
+/// {
+///     type Request = (hyper::Request, C);
+///     type Response = T::Response;
+///     type Error = T::Error;
+///     type Future = T::Future;
+///     fn call(&self, (req, context) : Self::Request) -> Self::Future {
+///         let (item, context) = context.pop();
+///         do_something_with_my_item(&item);
+///         self.inner.call((req, context))
+///     }
+/// }
+///
+/// # fn main() {}
+pub trait Pop<T> {
+    /// The type that remains after the value has been popped.
+    type Result;
+    /// Extracts a value.
+    fn pop(self) -> (T, Self::Result);
+}
 
-    /// Set the context logger
-    fn set_logger(&mut self, logger: slog::Logger);
+/// Defines a method for inserting a value, changing the resulting
+/// type. Used to specify that a hyper service adds some data from the context,
+/// making it available to later layers, e.g.
+///
+/// ```rust
+/// # extern crate hyper;
+/// # extern crate swagger;
+/// # extern crate futures;
+/// #
+/// # use swagger::context::*;
+/// # use futures::future::{Future, ok};
+/// # use std::marker::PhantomData;
+/// #
+/// # struct MyItem;
+/// #
+/// struct MiddlewareService<T, C> {
+///     inner: T,
+///     marker: PhantomData<C>,
+/// }
+///
+/// impl<T, C> hyper::server::Service for MiddlewareService<T, C>
+///     where
+///         C: Push<MyItem>,
+///         T: hyper::server::Service<Request = (hyper::Request, C::Result)>
+/// {
+///     type Request = (hyper::Request, C);
+///     type Response = T::Response;
+///     type Error = T::Error;
+///     type Future = T::Future;
+///     fn call(&self, (req, context) : Self::Request) -> Self::Future {
+///         let context = context.push(MyItem{});
+///         self.inner.call((req, context))
+///     }
+/// }
+///
+/// # fn main() {}
+pub trait Push<T> {
+    /// The type that results from adding an item.
+    type Result;
+    /// Inserts a value.
+    fn push(self, T) -> Self::Result;
 }
 
-impl HasLogger for Context {
-    fn logger(&self) -> &Option<slog::Logger> {
-        &self.logger
-    }
+/// Defines a struct that can be used to build up contexts recursively by
+/// adding one item to the context at a time, and a unit struct representing an
+/// empty context. The first argument is the name of the newly defined context struct
+/// that is used to add an item to the context, the second argument is the name of
+/// the empty context struct, and subsequent arguments are the types
+/// that can be stored in contexts built using these struct.
```

Comment:
Is there a reason for having to create a new empty context and name it? Why can't we just define a unit struct in this module and use that?

Reply:
I did have a reason, but in typing up the above comment I realised that it's not actually necessary.

The reason was so that the macro could implement traits defined in this crate on the empty context type when called in another crate (which requires that the crate calling the macro either owns  the trait or the type it's being implemented on. Since obviously the traits are defined here it would need to own the type). But the only trait the empty context needs to implement are `Default` and `Push<T>` for all `T`, and we can just define those impls here.

Reply:
Cool, sounds like that would be a nice improvement

Reply:
I just realised that this is needed after all, since the `Push<T>` impl refers to the cons-cell struct that's defined as part of the macro. 🙁